group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -936,62 +941,90 @@ private void append(
+                        // Otherwise, we write the current batch, allocate a new one and re-verify
+                        // whether the records fit in it.
+                        // If flushing fails, we don't catch the exception in order to let
+                        // the caller fail the current operation.
+                        flushCurrentBatch();
+                        maybeAllocateNewBatch(
+                            producerId,
+                            producerEpoch,
+                            verificationGuard,
+                            currentTimeMs
+                        );
+                    }
                 }
 
-                // Add the event to the list of pending events associated with the batch.
-                currentBatch.deferredEvents.add(event);
-
                 try {
-                    // Apply record to the state machine.
-                    if (replay) {
-                        for (int i = 0; i < records.size(); i++) {
-                            // We compute the offset of the record based on the last written offset. The
-                            // coordinator is the single writer to the underlying partition so we can
-                            // deduce it like this.
+                    for (int i = 0; i < records.size(); i++) {
+                        U recordToReplay = records.get(i);
+                        SimpleRecord recordToAppend = recordsToAppend.get(i);
+
+                        if (replay) {
                             coordinator.replay(
-                                currentBatch.nextOffset + i,
+                                currentBatch.nextOffset,
                                 producerId,
                                 producerEpoch,
-                                records.get(i)
+                                recordToReplay
                             );
                         }
-                    }
 
-                    // Append to the batch.
-                    for (SimpleRecord record : recordsToAppend) {
-                        currentBatch.builder.append(record);
+                        if (!isAtomic) {
+                            boolean hasRoomFor = currentBatch.builder.hasRoomFor(
+                                recordToAppend.timestamp(),
+                                recordToAppend.key(),
+                                recordToAppend.value(),
+                                recordToAppend.headers()
+                            );
+
+                            if (!hasRoomFor) {
+                                if (currentBatch.builder.numRecords() == 0) {
+                                    throw new RecordTooLargeException("Record " + recordToAppend + " in append to partition " + tp +
+                                        " exceeds exceeds the maximum configured size of " + currentBatch.maxBatchSize + ".");
+                                }
```

Comment:
Why do we need to check `numRecords`? If `hasRoomFor` is false, `numRecords` should be non zero https://github.com/apache/kafka/blob/206d0f809ad8438e9d89b7f731a8b995f952f0c0/clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java#L856-L858

Reply:
My understanding is that this was for when we can't fit any records into the batch because the record size is too large. !hasRoomFor means there is no more space in the batch for more records and numRecords == 0 means there are no records in the batch yet. Thus, the records are too big for any batch.

Reply:
I think I'm missing though what happens in the case we did have some records already in the builder but the record is too big for the batch.

Reply:
Oh, I forgot that `hasRoomFor` always accept one record. In this case, it will when the batch is written. I updated the code to reflect this. This is also covered by the test that I added.

Reply:
> I think I'm missing though what happens in the case we did have some records already in the builder but the record is too big for the batch.

The new test does exactly this.